File: src/main/ATOMExpression.java
package main;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class ATOMExpression extends ATOMElement {

    List<ATOMElement> children;

    public ATOMExpression() {
        children = new ArrayList<>();
    }

    public ATOMExpression(ATOMValue val) {
        this.children = Collections.singletonList(val);
    }

    public ATOMExpression(List<ATOMElement> children) {
        this.children = children;
    }

    private void setImplicitParentheticalScope() {
        List<ATOMOperation> implicitScopeOperations = Arrays.asList(
                ATOMOperation.FOREACH,
                ATOMOperation.IFOREACH,
                ATOMOperation.WHERE,
                ATOMOperation.IWHERE,
                ATOMOperation.MAP,
                ATOMOperation.INTO,
                ATOMOperation.THROUGH
        );
        for (int i=0;i<children.size();i++) {
            if (implicitScopeOperations.contains(children.get(i))) {
                if (i == children.size()-2 && children.get(children.size()-1) instanceof ATOMScope) {
                    return;
                }
                ATOMScope subScope = new ATOMScope();
                subScope.type = ATOMScopeType.PARENTHESIS;
                subScope.parent = parent;
                List<ATOMElement> implicitScope = new ArrayList<>();
                while (children.size()>i+1) {
                    ATOMElement child = children.remove(i+1);
                    child.parent = subScope;
                    implicitScope.add(child);
                }
                ATOMExpression subExpression = new ATOMExpression(implicitScope);
                subExpression.parent = subScope;
                subScope.components.add(subExpression);
                children.add(subScope);
                return;
            }
        }
    }

    private void setImplicitArrayLength() {
        for (int i=2;i<children.size();i++) {
            if (children.get(i) instanceof ATOMValueDynamic) {
                ATOMValueDynamic var = (ATOMValueDynamic) children.get(i);
                if (var.getType() == ATOMValueType.SCOPED) {
                    if (children.get(i-1) instanceof ATOMOperation && children.get(i-1)==ATOMOperation.DEREFERENCE) {
                        children.set(i, new ATOMValue(var.name));
                    }
                }
            }
        }
    }

    public ATOMValue eval() {
        setImplicitParentheticalScope();
        setImplicitArrayLength();
        List<ATOMElement> copyOfChildren = new ArrayList<>(children);
        while (copyOfChildren.size()>1) {
            int nextOperationIndex = -1;
            for (int i=0;i<copyOfChildren.size();i++) {
                if (copyOfChildren.get(i) instanceof ATOMOperation) {
                    if (nextOperationIndex==-1) {
                        nextOperationIndex = i;
                    }
                    if (((ATOMOperation) copyOfChildren.get(i)).order < ((ATOMOperation) copyOfChildren.get(nextOperationIndex)).order) {
                        nextOperationIndex = i;
                    }
                    // 100+ indicates a right to left order of operations
                    if (((ATOMOperation) copyOfChildren.get(i)).order == ((ATOMOperation) copyOfChildren.get(nextOperationIndex)).order
                            && ((ATOMOperation) copyOfChildren.get(i)).order>=100) {
                        nextOperationIndex = i;
                    }
                }
            }

            if (nextOperationIndex==-1) {
                throw new RuntimeException("Probably a syntax error somewhere, not enough operators:\n" + this.children);
            }

            ATOMElement left=null, right=null;
            if (nextOperationIndex<copyOfChildren.size()-1 && !(copyOfChildren.get(nextOperationIndex+1) instanceof ATOMOperation)) {
                right = copyOfChildren.get(nextOperationIndex+1);
                copyOfChildren.remove(nextOperationIndex+1);
            }
            ATOMOperation op = (ATOMOperation) copyOfChildren.remove(nextOperationIndex);
            if (nextOperationIndex>0 && !(copyOfChildren.get(nextOperationIndex-1) instanceof ATOMOperation)) {
                left = copyOfChildren.get(nextOperationIndex-1);
                copyOfChildren.remove(nextOperationIndex-1);
            }
            ATOMValue newVal = op.eval(left, right);
            if (left == null) {
                copyOfChildren.add(nextOperationIndex, newVal);
            } else {
                copyOfChildren.add(nextOperationIndex - 1, newVal);
            }
        }
        if (copyOfChildren.size()==0 || copyOfChildren.get(0)==null) {
            return null;
        }
        return copyOfChildren.get(0).eval();
    }

    public ATOMValue compute() {
        if (children.size()==1) {
            return children.get(0).compute();
        }
        return eval();
    }

//     BUNDLER: BEGIN IGNORE
    public String toString() {
        List<String> childrenStr = children.stream().map(ATOMElement::toString).collect(Collectors.toList());
        return String.join(" ", childrenStr);
    }
    // BUNDLER: END IGNORE
    // BUNDLER: BEGIN INJECT
//    public String toString() {
//        String toReturn = "";
//        for (ATOMElement elem:children) {
//            toReturn += elem.toString() + " ";
//        }
//        return toReturn;
//    }
    // BUNDLER: END INJECT

}

File: web_interpreter/atom/ATOMExpression.js
import ATOMElement from './ATOMElement.js';
import ATOMOperation from './ATOMOperation.js';
import ATOMScope from './ATOMScope.js';
import ATOMValue from './ATOMValue.js';
import ATOMValueDynamic from './ATOMValueDynamic.js';

export default class ATOMExpression {

  constructor(arg) {
    this.children = null;
    if (arg instanceof Array) {
      this.children = arg;
      return;
    }
    if (arg) {
      this.children = [arg];
      return;
    }
    this.children = [];
  }

  setImplicitParentheticalScope() {
    let implicitScopeOperations = [
      ATOMOperation.FOREACH,
      ATOMOperation.IFOREACH,
      ATOMOperation.WHERE,
      ATOMOperation.IWHERE,
      ATOMOperation.MAP,
      ATOMOperation.INTO,
      ATOMOperation.THROUGH
    ];
    for (let i=0;i<this.children.length;i++) {
      if (implicitScopeOperations.indexOf(this.children[i])>=0) {
        if (i == this.children.length-2 && this.children[this.children.length-1] instanceof ATOMScope) {
          return;
        }
        let subScope = new ATOMScope();
        subScope.type = 'PARENTHESIS';
        subScope.parent = this.parent;
        let implicitScope = [];
        while (this.children.length>i+1) {
          let child = this.children.splice(i+1, 1)[0];
          child.parent = subScope;
          implicitScope.push(child);
        }
        let subExpression = new ATOMExpression(implicitScope);
        subExpression.parent = subScope;
        subScope.components.push(subExpression);
        this.children.add(subScope);
        return;
      }
    }
  }

  setImplicitArrayLength() {
    for (let i=2;i<this.children.length;i++) {
      if (this.children[i] instanceof ATOMValueDynamic) {
        let val = this.children[i];
        if (val.getType()=='SCOPED') {
          if (this.children[i-1] instanceof ATOMOperation && children[i-1]==ATOMOperation.DEREFERENCE) {
            children[i] = new ATOMValue(val.name);
          }
        }
      }
    }
  }

  eval() {
    this.setImplicitParentheticalScope();
    this.setImplicitArrayLength();
    let copyOfChildren = [...this.children];
    while (copyOfChildren.length>1) {
      let nextOperationIndex = -1;
      for (let i=0;i<copyOfChildren.length;i++) {
        if (copyOfChildren[i] instanceof ATOMOperation) {
          if (nextOperationIndex==-1) {
            nextOperationIndex = i;
          }
          if (copyOfChildren[i].order < copyOfChildren[nextOperationIndex].order
            || copyOfChildren[i].order == copyOfChildren[nextOperationIndex].order
            && copyOfChildren[i].order == ATOMOperation.ORDER_UNARY) {
            nextOperationIndex = i;
          }
          if (copyOfChildren[i].order == copyOfChildren[nextOperationIndex].order && copyOfChildren[i].order>=100) {
            nextOperationIndex = i;
          }
        }
      }
      if (nextOperationIndex==-1) {
        throw 'Probably a syntax error somewhere, not enough operators:\n' + this.children.toString();
      }
      let left = null, right = null;
      if (nextOperationIndex<copyOfChildren.length-1 && !(copyOfChildren[nextOperationIndex+1] instanceof ATOMOperation)) {
        right = copyOfChildren[nextOperationIndex+1];
        copyOfChildren.splice(nextOperationIndex+1, 1);
      }
      let op = copyOfChildren.splice(nextOperationIndex, 1)[0];
      if (nextOperationIndex>0 && !(copyOfChildren[nextOperationIndex-1] instanceof ATOMOperation)) {
        left = copyOfChildren[nextOperationIndex-1];
        copyOfChildren.splice(nextOperationIndex-1, 1);
      }
      let newVal = op.eval(left, right);
      if (left == null) {
        copyOfChildren.splice(nextOperationIndex, 0, newVal);
      } else {
        copyOfChildren.splice(nextOperationIndex - 1, 0, newVal);
      }
    }
    if (copyOfChildren.length==0 || copyOfChildren[0]==null) {
      return null;
    }
    return copyOfChildren[0].eval();
  }

  compute() {
    if (this.children.length==1) {
      return this.children[0].compute();
    }
    return this.eval();
  }

  toString() {
    return this.children.map(c => c.toString()).join(' ');
  }

}